### PR TITLE
Now all together: Perf and CIDs

### PR DESCRIPTION
**Pull request by dignifiedquire**
This is the perf improvement PR #83 and with the cid changes from #76 rebased onto them. All tests are passing and benchmarks are close to what I had before.

### CODE REVIEW

**Comment by dignifiedquire** [comment]
This is ready for review, speed is now even a bit better than on #83 with some more optimizations.

**Comment by diasdavid** [comment]
For this to be merged all the way in, we want to have interop bitswap tests in js-ipfs API, which creates a tree of dependencies

- [ ] Merging ipfs-daemon features into ipfsd-ctl (https://github.com/ipfs/js-ipfsd-ctl/issues/136)
- [ ] Adding the interop tests with the newly ipfsd-ctl https://github.com/ipfs/js-ipfs/pull/674
- [ ] Testing this whole thing through
- [ ] Merge and release https://github.com/ipfs/js-ipfs-block-service/pull/51
- [ ] Merge and release this PR
- [ ] Update deps in js-ipfs
- [ ] 🚢 🚀

File: README.md [hunk 1]
```diff
@@ -62,6 +62,114 @@ Loading this module through a script tag will make the `IpfsBitswap` object avai
...
+#### `cancelWants(cids)`
+
+- `cid: CID|[]CID`
+
+Cancel previously requested keys.
+
+#### `putStream()`
+
+Returns a duplex `pull-stream` that emits an object `{key: Multihash}` for every written block when it was stored.
+Objects passed into here should be of the form `{data: Buffer, key: Multihash}`
```
Comment (diasdavid) [inline README.md]: It is no long `data: Buffer, key Multihash`, instead it is `block: block, cid: cid`

File: README.md [hunk 2]
```diff
@@ -62,6 +62,114 @@ Loading this module through a script tag will make the `IpfsBitswap` object avai
...
+
+#### `put(blockAndCid, callback)`
+
+- `blockAndKey: {data: Buffer, cid: CID}`
```
Comment (diasdavid) [inline README.md]: `s/data: Buffer/block: Block`

File: benchmarks/index.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      cb()
+    })
+  })
+}
+
+function createBlocks (n, blockFactor) {
+  return _.map(_.range(n * blockFactor), () => {
+    return new Block(crypto.randomBytes(n * blockFactor))
+  })
+}
```
Comment (diasdavid) [inline benchmarks/index.js]: ❤️ benchmarks!

File: package.json [hunk 4]
```diff
@@ -54,20 +56,26 @@
...
     "lodash.debounce": "^4.0.8",
+    "lodash.find": "^4.6.0",
+    "lodash.groupby": "^4.6.0",
     "lodash.isequalwith": "^4.4.0",
     "lodash.isundefined": "^3.0.1",
-    "multihashes": "^0.3.0",
+    "lodash.pullallwith": "^4.7.0",
+    "lodash.uniqwith": "^4.5.0",
+    "lodash.values": "^4.3.0",
+    "multihashes": "^0.3.1",
```
Comment (diasdavid) [inline package.json]: Not used anymore

File: src/components/decision-engine/index.js [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    if (!cids.length) {
+      return
+    }
+    // Check all connected peers if they want the block we received
+    for (let l of this.ledgerMap.values()) {
+      cids
+        .map((k) => l.wantlistContains(k))
+        .filter(Boolean)
+        .forEach((e) => {
+          // this.peerRequestQueue.push(e, l.partner)
```
Comment (diasdavid) [inline src/components/decision-engine/index.js]: I believe peerRequestQueue are no longer a thing, right?

File: src/components/network/index.js [hunk 6]
```diff
@@ -75,17 +91,27 @@ module.exports = class Network {
   }
 
   _onPeerMux (peerInfo) {
+    if (!this._running) {
+      return
+    }
```
Comment (diasdavid) [inline src/components/network/index.js]: good catch :)

File: src/components/decision-engine/index.js (hunk 5 above)
Comment (diasdavid) [inline src/components/decision-engine/index.js]: Did you ripped out PeerRequestQueue and MessageQueue, or just PeerRequestQueue? From your comment in https://github.com/ipfs/js-ipfs-bitswap/issues/88#issuecomment-268521978, it sounds like you have removed both, but I believe only PeerRequestQueue is gone.

Comment (dignifiedquire) [inline src/components/decision-engine/index.js]: Which is important to keep on a per peer basis, as we send messages out on a per peer basis.

Comment (diasdavid) [inline src/components/decision-engine/index.js]: I need to update the graph :)

Comment (dignifiedquire) [inline src/components/decision-engine/index.js]: also feel free to suggest a better name :)

File: README.md (hunk 2 above)
Comment (dignifiedquire) [inline README.md]: fixed

File: README.md (hunk 1 above)
Comment (dignifiedquire) [inline README.md]: fixed

File: package.json (hunk 4 above)
Comment (dignifiedquire) [inline package.json]: fixed

**Comment by dignifiedquire** [comment]
@diasdavid linking this together with the other repo PR, and the newly created interop tests and this adjustment in js-ipfs

```diff
diff --git a/src/core/components/go-online.js b/src/core/components/go-online.js
index 793a026..28894f7 100644
--- a/src/core/components/go-online.js
+++ b/src/core/components/go-online.js
@@ -14,7 +14,6 @@ module.exports = function goOnline (self) {
       }

       self._bitswap = new Bitswap(
-        self._libp2pNode.peerInfo,
         self._libp2pNode,
         self._repo.blockstore,
         self._libp2pNode.peerBook
(END)
```

 I can confirm we can transfer data between go and js using bitswap 1.1.0 🎉

**Comment by diasdavid** [comment]
> 14:06 <@dignifiedquire> daviddias: https://github.com/ipfs/js-ipfs-bitswap/pull/84#issuecomment-268806034 🎉 bitswap is ready to rock
14:06 <@dignifiedquire> nice job! worked out of th box

AWESOME :D 

Thank you :)

**Comment by dignifiedquire** [comment]
:tada:

**Pull request merged**